### PR TITLE
Allow different environments with different set of plugins for testing features

### PR DESCRIPTION
**Pull request by johmsalas**
@peterfication. Some of these files are not even used, but they kind of describe the design for splitting tests according to different configuration of plugins

Minimal is working fine, and will start adding the next set of configurations in different PRs.

It is ready to be reviewed now

### CODE REVIEW

File: tests/run.sh [hunk 1]
```diff
@@ -1,3 +1,13 @@
 #!/bin/sh
 
-nvim --headless -u tests/init.lua -c "PlenaryBustedDirectory tests/ {minimal_init = 'tests/init.lua', sequential = true}"
+# The Tests Runner runs tests for all the environments
```
Comment (johmsalas) [inline tests/run.sh:3]: @peterfication this is a draft, the idea is to clone the repos of plugins and use what is required isolated according to the tested features, like Telescope

File: tests/run.sh [hunk 2]
```diff
@@ -1,3 +1,13 @@
...
+
+mkdir -p ~/.local/share/nvim/site/pack/vendor/start
+git clone --depth 1 https://github.com/nvim-lua/plenary.nvim ~/.local/share/nvim/site/pack/vendor/start/plenary.nvim
+git clone --depth 1 https://github.com/nvim-telescope/telescope.nvim ~/.local/share/nvim/site/pack/vendor/start/telescope.nvim
+
+nvim --headless -u tests/environments/minimal.vim -c "PlenaryBustedDirectory tests/ {minimal_init = 'tests/environments/minimal.vim', sequential = true}"
```
Comment (johmsalas) [inline tests/run.sh:9]: The tests runner would run everything, and later we can parametrized to be able to run just a subset

File: tests/init.lua [hunk 3]
```diff
@@ -1,35 +0,0 @@
...
-
-function M.root(root)
-  local f = debug.getinfo(1, "S").source:sub(2)
-  return vim.fn.fnamemodify(f, ":p:h:h") .. "/" .. (root or "")
-end
-
----@param plugin string
-function M.load(plugin)
-  local name = plugin:match(".*/(.*)")
-  local package_root = M.root(".tests/site/pack/deps/start/")
```
Comment (johmsalas) [inline tests/init.lua:13]: I'll check where this is getting cloned to avoid populating usable paths here like ~/.local/share/nvim/site/pack/vendor/start/

**Comment by johmsalas** [comment]
Actually, using vim files might not be a good idea, it could use Lua to have access to the feature per Neovim version table https://github.com/johmsalas/text-case.nvim/issues/62#issuecomment-1817410150

File: tests/run.sh (hunk 1 above)
Comment (peterfication) [reply tests/run.sh:3]: We probably need some kind of vim setup inside a spec file then, right?

File: lua/constants.lua [hunk 4]
```diff
@@ -0,0 +1,13 @@
+local is_feature_available = function(feature)
```
Comment (johmsalas) [inline lua/constants.lua:1]: Just an initial idea, it is not used yet

File: tests/run.sh [hunk 5]
```diff
@@ -1,3 +1,46 @@
...
+  fi
+}
+
+mkdir -p $TEST_MINIMAL_DIR
+clone $GITHUB_PLENARY "$TEST_MINIMAL_DIR/plenary.nvim"
+nvim --headless -u tests/environments/minimal.lua -c "PlenaryBustedDirectory tests/ {minimal_init = 'tests/environments/minimal.lua', sequential = true}"
+
+# mkdir -p $TEST_LSP_DIR
+# git clone --depth 1 $GITHUB_PLENARY "$TEST_LSP_DIR/plenary.nvim"
+# nvim --headless -u tests/environments/lsp.lua -c "PlenaryBustedDirectory tests/ {minimal_init = 'tests/environments/lsp.lua', sequential = true}"
```
Comment (peterfication) [inline tests/run.sh:35]: Is the idea then to split by directory?

File: tests/environments/test-env-helpers.lua
Comment (peterfication) [inline tests/environments/test-env-helpers.lua:1]: Is this file needed?

**Review by peterfication — APPROVED** [review]
Looks like a good approach to me 👌😊

I would squash some of the commits. You've merged main already, then a squashing is not possible anymore. So would get rid of the merge commit, rebase against main and then squash the commits with the same commit message.

File: tests/run.sh (hunk 5 above)
Comment (johmsalas) [reply tests/run.sh:35]: The environments are split into directories, I'm considering 2 approaches for splitting the tests:

* Run only the tests for the plugin environment. ie: 
```
nvim -u tests/environments/lsp.lua -c "PlenaryBustedDirectory tests/lsp/
nvim -u tests/environments/telescope.lua -c "PlenaryBustedDirectory tests/telescope/
```

* Run all the tests for all the plugin environments
The tests would contain guard statements to avoid including tests according to the feature availability

File: tests/environments/test-env-helpers.lua
Comment (johmsalas) [reply tests/environments/test-env-helpers.lua:1]: This is duplicated by all the plugin environment init files. I couldn't easily require it, so I went with a "duplicate" code approach. Ideally this file will be used

**Comment by johmsalas** [comment]
@peterfication Is it too bad if this get merged as it is?
btw, I do see enabled the squash and merge option

Any online runbook you suggest me to improve my git collaboration?

**Comment by johmsalas** [comment]
Maybe this one https://docs.gitlab.com/ee/topics/git/git_rebase.html 

Just, I'd prefer not to block the PR while I figure it out what was wrong with this one
image

File: tests/environments/minimal.lua [hunk 6]
```diff
@@ -0,0 +1,17 @@
+local M = {}
```
Comment (peterfication) [inline tests/environments/minimal.lua:1]: This would work:

```lua
local environments_folder_path = vim.fn.fnamemodify(debug.getinfo(1, "S").source:sub(2), ":p:h") .. "/"
local helpers = dofile(environments_folder_path .. "test-env-helpers.lua")

helpers.setup(".tests/minimal/site")
```

File: tests/environments/test-env-helpers.lua
Comment (peterfication) [reply tests/environments/test-env-helpers.lua:1]: I added a comment to make it work at the tests/environments/minimal.lua file

**Comment by peterfication** [comment]
> Maybe this one https://docs.gitlab.com/ee/topics/git/git_rebase.html
> 
> Just, I'd prefer not to block the PR while I figure it out what was wrong with this one

Yes, this guide is a good starting point. For the next PR then :)
